source/extensions/network/dns_resolver/cares/dns_impl.cc

```diff
@@ -208,6 +215,11 @@ void DnsResolverImpl::AddrInfoPendingResolution::onAresGetAddrInfoCallback(
 
     ASSERT(addrinfo != nullptr);
     ares_freeaddrinfo(addrinfo);
+  } else if (isResponseWithNoRecords(status)) {
+    // Treat `ARES_ENODATA` or `ARES_ENOTFOUND` here as success to populate back the
+    // "empty records" response.
+    pending_response_.status_ = ResolutionStatus::Success;
+    ASSERT(addrinfo == nullptr);
```

Comment:
Are we sure the caller always set addrinfo to be nullptr when the status is ARES_ENODATA or ARES_ENOTFOUND?

Others LGTM.

Reply:
I think so, otherwise the current DnsImpl should have been leaking memory in the non-success callback branch, c-ares [doc](https://c-ares.org/ares_getaddrinfo.html) and [code](https://github.com/c-ares/c-ares/blob/main/src/lib/ares_getaddrinfo.c#L399) for reference.